network/connPerfMon.go

```diff
@@ -298,31 +305,58 @@ func (pm *connectionPerformanceMonitor) updateMessageIdlingInterval(now int64) (
 
 func (pm *connectionPerformanceMonitor) pruneOldMessages(now int64) {
 	oldestMessage := now - int64(pmMaxMessageWaitTime)
-	for digest, pendingMsg := range pm.pendingMessages {
-		if oldestMessage < pendingMsg.firstPeerTime {
+	prunedBucketsCount := 0
+	for bucketIdx, currentMsgBucket := range pm.pendingMessagesBuckets {
+		if currentMsgBucket.endTime > oldestMessage {
+			pm.pendingMessagesBuckets[bucketIdx-prunedBucketsCount] = currentMsgBucket
 			continue
 		}
-		for peer := range pm.monitoredConnections {
-			if msgTime, hasPeer := pendingMsg.peerMsgTime[peer]; hasPeer {
-				msgDelayInterval := msgTime - pendingMsg.firstPeerTime
-				pm.connectionDelay[peer] += msgDelayInterval
-			} else {
-				// we never received this message from this peer.
-				pm.connectionDelay[peer] += int64(pmUndeliveredMessagePenaltyTime)
+		for _, pendingMsg := range currentMsgBucket.messages {
+			for peer := range pm.monitoredConnections {
+				if msgTime, hasPeer := pendingMsg.peerMsgTime[peer]; hasPeer {
+					msgDelayInterval := msgTime - pendingMsg.firstPeerTime
+					pm.connectionDelay[peer] += msgDelayInterval
+				} else {
+					// we never received this message from this peer.
+					pm.connectionDelay[peer] += int64(pmUndeliveredMessagePenaltyTime)
+				}
 			}
 		}
-		delete(pm.pendingMessages, digest)
+		prunedBucketsCount++
 	}
+	pm.pendingMessagesBuckets = pm.pendingMessagesBuckets[:len(pm.pendingMessagesBuckets)-prunedBucketsCount]
 }
 
 func (pm *connectionPerformanceMonitor) accumulateMessage(msg *IncomingMessage, newMessages bool) {
 	msgDigest := generateMessageDigest(msg.Tag, msg.Data)
 
-	pendingMsg := pm.pendingMessages[msgDigest]
+	var msgBucket *pmPendingMessageBucket
+	var pendingMsg *pmMessage
+	var msgFound bool
+	// try to find the message. It's more likely to be found in the most recent bucket, so start there and go backward.
+	for bucketIndex := range pm.pendingMessagesBuckets {
+		currentMsgBucket := pm.pendingMessagesBuckets[len(pm.pendingMessagesBuckets)-1-bucketIndex]
+		if pendingMsg, msgFound = currentMsgBucket.messages[msgDigest]; msgFound {
+			msgBucket = currentMsgBucket
+			break
+		}
+		if msg.Received >= currentMsgBucket.startTime && msg.Received <= currentMsgBucket.endTime {
+			msgBucket = currentMsgBucket
+		}
+	}
 	if pendingMsg == nil {
 		if newMessages {
+			if msgBucket == nil {
+				// no bucket was found. create one.
+				msgBucket = &pmPendingMessageBucket{
+					messages:  make(map[crypto.Digest]*pmMessage),
+					startTime: msg.Received - (msg.Received % int64(pmMessageBucketDuration)), // align with pmMessageBucketDuration
+				}
+				msgBucket.endTime = msgBucket.startTime + int64(pmMessageBucketDuration) - 1
```

Comment:
nit: seems like endTime can be
```
msgBucket.endTime = msgBucket.startTime + int64(pmMessageBucketDuration
```
if use semi-open interval above:
```
if msg.Received >= currentMsgBucket.startTime && msg.Received < currentMsgBucket.endTime {
```

Reply:
maybe I'm missing something, but I don't see any particular "value" in that. i.e. the compiler would already perform the "-1" as part of the computation, and doing a comparison operation using < vs <= has the same performance characteristics. I can see that there might be some subtle value in dropping the endTime completely, as it can always be calculated off the startTime, but again - I don't think we will ever see any concrete benefits doing so.